### PR TITLE
scripts: dts: Replace 'sub-node:' with more general 'child-binding:'

### PR DESCRIPTION
**Pull request by ulfalizer**
Deprecate 'sub-node:' and add a more general 'child-binding:' mechanism
to bindings. Keep supporting 'sub-node:', but print a deprecation
warning when it's used.

Like 'sub-node:', 'child-binding:' gives a binding to child nodes, but
the binding is required to be a complete binding, and is treated (and
checked) like a normal binding.

'child-binding:' can in turn contain another 'child-binding:', up to any
number of levels. This is automatic from treating it like a normal
binding, and from the code initializing parent Devices before child
Devices.

This lets nodes give bindings to grandchildren.

For example, take this devicetree fragment:

    parent {
            compatible = "foo";
            child-1 {
                    grandchild-1 {
                            ...
                    };
                    grandchild-2 {
                            ...
                    };
            };
            child-2 {
                    grandchild-3 {
                            ...
                    };
            };
    };

The binding for 'foo' could provide bindings for grandchild-1/2/3 like
this:

    compatible: "foo"

    # Binding for children
    child-binding:
        title: ...
        description: ...

        ...

        # Binding for grandchildren
        child-binding:
            title: ...
            description: ...

            properties:
                ...

Due to implementation issues with the old devicetree scripts, only two
levels of 'child-binding:' is supported for now. This limitation will go
away in Zephyr 2.2.

Piggyback shortening 'description:' and 'title:' in some bindings that
provide child bindings. This makes the generated header a bit neater.

### CODE REVIEW

**Comment by ulfalizer** [comment]
@b0661 wanted this for some PINMUX stuff I think. It's nice to have a general mechanism in any case.

**Comment by ulfalizer** [comment]
@galak 
Ping

**Review by b0661 — APPROVED** [review]
LGTM

File: dts/binding-template.yaml [hunk 1]
```diff
@@ -207,6 +195,66 @@ properties:
...
+# parent {
+#         compatible = "foo";
+#         child {
+#                 grandchild {
+#                         prop = <123>;
+#                 };
+#         };
+# }
+#
+# WARNING: Due to implementation issues with legacy code, only up to two levels
```
Comment (galak) [inline dts/binding-template.yaml]: Just curious, is this limitation enforced or just something about how the old code works?

**Comment by galak** [comment]
If you haven't already, please update #19197 with regards to this change.

**Review by galak — APPROVED** [review]

File: dts/binding-template.yaml (hunk 1 above)
Comment (ulfalizer) [reply dts/binding-template.yaml]: Not enforced now. The grandgrandchildren will just silently not get the binding in the old code if you try to nest it to three levels.

Could add a check for it, though I wonder if we will ever have three levels in practice.

**Pull request merged**